studiopress/pattern-manager
# Replace the `save-pattern` endpoint with filtering in PHP

* PR [40](https://github.com/studiopress/pattern-manager/pull/40) left some [known issues](https://github.com/studiopress/pattern-manager/pull/40#issue-1563099080) that don't have easy fixes in JS.
* This fixes both of them.
* It lazily creates a `pm_pattern` post when editing a pattern. Patterns you never edit will never have a post, and each pattern should only have 1 post.
* The only thing it sets in the post is the `->post_name`.
* It doesn't store the pattern in `post_content`, it gets that via the pattern `.php` file and adds it to `post_content` via a filter.
* Similar with the post meta, like categories and keywords.
* It gets those from the pattern `.php` file.
* This is hacky, but I think better than my approach in #40 
* It's much easier to extend WP in PHP filters and actions. Extending Gutenberg is really hard.
* It works with the block editor, which requires a post ID.
* We give it a post ID, but we do the minimum possible with that post.





### How to test
There's still work left

1. Click 'Create New Pattern'
2. Make some edits
3. Click Command S
<img width="1004" alt="Screenshot 2023-02-09 at 10 39 35 AM" src="https://user-images.githubusercontent.com/4063887/217879764-51a16b0e-4847-40c7-848d-dd6c13b60a61.png">

4. Reload the browser
5. Expected: There's no warning 'Are you sure you want to leave the page', like before
6. Expected: The edits persist
7. Go to the PHP file for the pattern
8. Edit the pattern content and meta values, like Keywords and Post Types 
9. Reload the browser
10. Expected: The edits from the .php file appear, as the .php file is still the source of truth: <img width="896" alt="Screenshot 2023-02-09 at 10 45 09 AM" src="https://user-images.githubusercontent.com/4063887/217881266-22672922-9d64-4698-b953-9363b56e36b2.png">
